File: frontend/src/shared/services/answerService.ts
// Answer 도메인 API 서비스
import { api, handleApiError } from './apiClient';

// ============================================================================
// 타입 정의 (백엔드 Pydantic 스키마와 일치)
// ============================================================================

/**
 * 단일 답변 업데이트 요청 스키마
 */
export interface AnswerUpdatePayload {
  requirement_id: string;
  answer_data: any; // 다양한 타입의 답변 데이터 (string, number, boolean, object 등)
}

/**
 * 배치 답변 업데이트 요청 스키마
 */
export interface AnswerBatchUpdatePayload {
  answers: AnswerUpdatePayload[];
}

/**
 * 배치 답변 처리 응답 스키마
 */
export interface AnswerBatchResponse {
  success: boolean;
  message: string;
  processed_count: number;
  created_count: number;
  updated_count: number;
  failed_count: number;
  failed_requirements: string[];
}

/**
 * 답변 응답 스키마
 */
export interface AnswerResponse {
  answer_id: string;
  user_id: string;
  requirement_id: string;
  answer_value_text?: string | null;
  answer_value_number?: number | null;
  answer_value_boolean?: boolean | null;
  answer_value_date?: string | null; // ISO 날짜 문자열
  answer_value_json?: any | null;
  answered_at: string; // ISO 날짜 문자열
  last_edited_at: string; // ISO 날짜 문자열
}

// ============================================================================
// API 서비스 함수들
// ============================================================================

/**
 * 답변 배치 저장/업데이트 (UPSERT)
 * 
 * @param answers 저장할 답변 목록
 * @returns 처리 결과
 */
export const saveAnswersBatch = async (
  answers: AnswerUpdatePayload[]
): Promise<AnswerBatchResponse> => {
  try {
    console.log('🚀 답변 배치 저장 요청:', { count: answers.length });
    
    const payload: AnswerBatchUpdatePayload = { answers };
    
    const response = await api.post<AnswerBatchResponse>(
      '/api/disclosure/answers/batch',
      payload
    );
    
    console.log('✅ 답변 배치 저장 성공:', response);
    return response;
    
  } catch (error: any) {
    console.error('❌ 답변 배치 저장 실패:', error);
    const apiError = handleApiError(error);
    throw apiError;
  }
};

/**
 * 내 답변 목록 조회
 * 
 * @returns 사용자의 모든 답변 목록
 */
export const getMyAnswers = async (): Promise<AnswerResponse[]> => {
  try {
    console.log('🚀 내 답변 목록 조회 요청');
    
    const response = await api.get<AnswerResponse[]>(
      '/api/disclosure/answers/my'
    );
    
    console.log('✅ 내 답변 목록 조회 성공:', { count: response.length });
    return response;
    
  } catch (error: any) {
    console.error('❌ 내 답변 목록 조회 실패:', error);
    const apiError = handleApiError(error);
    throw apiError;
  }
};

/**
 * 특정 요구사항에 대한 내 답변 조회
 * 
 * @param requirementId 요구사항 ID
 * @returns 해당 요구사항에 대한 답변
 */
export const getMyAnswerByRequirement = async (
  requirementId: string
): Promise<AnswerResponse> => {
  try {
    console.log('🚀 특정 답변 조회 요청:', { requirementId });
    
    const response = await api.get<AnswerResponse>(
      `/api/disclosure/answers/my/${requirementId}`
    );
    
    console.log('✅ 특정 답변 조회 성공:', response);
    return response;
    
  } catch (error: any) {
    console.error('❌ 특정 답변 조회 실패:', error);
    const apiError = handleApiError(error);
    throw apiError;
  }
};

/**
 * 특정 요구사항에 대한 내 답변 삭제
 * 
 * @param requirementId 요구사항 ID
 * @returns 삭제 결과 메시지
 */
export const deleteMyAnswer = async (
  requirementId: string
): Promise<{ message: string }> => {
  try {
    console.log('🚀 답변 삭제 요청:', { requirementId });
    
    const response = await api.delete<{ message: string }>(
      `/api/disclosure/answers/my/${requirementId}`
    );
    
    console.log('✅ 답변 삭제 성공:', response);
    return response;
    
  } catch (error: any) {
    console.error('❌ 답변 삭제 실패:', error);
    const apiError = handleApiError(error);
    throw apiError;
  }
};

// ============================================================================
// 유틸리티 함수들
// ============================================================================

/**
 * 답변 데이터 유효성 검증
 * 
 * @param answer 검증할 답변 데이터
 * @returns 유효성 검증 결과
 */
export const validateAnswerPayload = (answer: AnswerUpdatePayload): {
  isValid: boolean;
  errors: string[];
} => {
  const errors: string[] = [];
  
  if (!answer.requirement_id || answer.requirement_id.trim() === '') {
    errors.push('requirement_id는 필수입니다.');
  }
  
  if (answer.answer_data === undefined || answer.answer_data === null) {
    errors.push('answer_data는 필수입니다.');
  }
  
  return {
    isValid: errors.length === 0,
    errors,
  };
};

/**
 * 배치 답변 데이터 유효성 검증
 * 
 * @param answers 검증할 답변 목록
 * @returns 유효성 검증 결과
 */
export const validateAnswersBatch = (answers: AnswerUpdatePayload[]): {
  isValid: boolean;
  errors: string[];
} => {
  const errors: string[] = [];
  
  if (!Array.isArray(answers) || answers.length === 0) {
    errors.push('답변 목록이 비어있습니다.');
    return { isValid: false, errors };
  }
  
  answers.forEach((answer, index) => {
    const validation = validateAnswerPayload(answer);
    if (!validation.isValid) {
      validation.errors.forEach(error => {
        errors.push(`답변 ${index + 1}: ${error}`);
      });
    }
  });
  
  return {
    isValid: errors.length === 0,
    errors,
  };
};

// ============================================================================
// Export
// ============================================================================

export default {
  saveAnswersBatch,
  getMyAnswers,
  getMyAnswerByRequirement,
  deleteMyAnswer,
  validateAnswerPayload,
  validateAnswersBatch,
}; 

File: disclosure-service/app/domain/service/answer_service.py
# Answer 도메인 Service
import logging
import json
from typing import Dict, Any, List, Optional
from uuid import UUID
from datetime import datetime
from sqlalchemy.orm import Session

from app.domain.repository.answer_repository import AnswerRepository
from app.domain.repository.disclosure_repository import DisclosureRepository
from app.domain.model.answer_schema import (
    AnswerBatchUpdatePayload, 
    AnswerBatchResponse,
    AnswerResponse
)

logger = logging.getLogger("disclosure-service")

class AnswerService:
    """Answer 도메인 비즈니스 로직 처리 계층"""
    
    def __init__(self, db: Session, answer_repository: AnswerRepository, disclosure_repository: DisclosureRepository):
        self.db = db
        self.answer_repository = answer_repository
        self.disclosure_repository = disclosure_repository
    
    def _map_answer_data_to_columns(self, answer_data: Any, data_required_type: str) -> Dict[str, Any]:
        """
        답변 데이터를 data_required_type에 따라 적절한 컬럼에 매핑합니다.
        
        Args:
            answer_data: 사용자가 입력한 답변 데이터
            data_required_type: 요구사항의 데이터 타입
            
        Returns:
            Dict[str, Any]: 데이터베이스 컬럼에 매핑된 데이터
        """
        column_data = {
            "answer_value_text": None,
            "answer_value_number": None,
            "answer_value_boolean": None,
            "answer_value_date": None,
            "answer_value_json": None,
            "last_edited_at": datetime.utcnow()
        }
        
        try:
            if data_required_type in ["text", "textarea", "select"]:
                column_data["answer_value_text"] = str(answer_data) if answer_data is not None else None
                
            elif data_required_type in ["number", "integer", "float"]:
                if answer_data is not None:
                    column_data["answer_value_number"] = float(answer_data)
                    
            elif data_required_type == "boolean":
                if answer_data is not None:
                    if isinstance(answer_data, bool):
                        column_data["answer_value_boolean"] = answer_data
                    elif isinstance(answer_data, str):
                        column_data["answer_value_boolean"] = answer_data.lower() in ["true", "yes", "1"]
                    else:
                        column_data["answer_value_boolean"] = bool(answer_data)
                        
            elif data_required_type == "date":
                if answer_data is not None:
                    if isinstance(answer_data, str):
                        column_data["answer_value_date"] = datetime.fromisoformat(answer_data)
                    elif isinstance(answer_data, datetime):
                        column_data["answer_value_date"] = answer_data
                        
            elif data_required_type in [
                "multiselect", "checkbox", "structured_list", "table_input", 
                "json", "object", "array"
            ]:
                if answer_data is not None:
                    # 이미 dict/list인 경우 그대로, 문자열인 경우 JSON 파싱 시도
                    if isinstance(answer_data, (dict, list)):
                        column_data["answer_value_json"] = answer_data
                    else:
                        try:
                            column_data["answer_value_json"] = json.loads(str(answer_data))
                        except json.JSONDecodeError:
                            # JSON 파싱 실패 시 텍스트로 저장
                            column_data["answer_value_text"] = str(answer_data)
                            
            else:
                # 알 수 없는 타입은 JSON으로 저장 시도, 실패 시 텍스트로 저장
                if answer_data is not None:
                    if isinstance(answer_data, (dict, list)):
                        column_data["answer_value_json"] = answer_data
                    else:
                        column_data["answer_value_text"] = str(answer_data)
                        
        except Exception as e:
            logger.warning(f"데이터 매핑 중 오류 발생, 텍스트로 저장: {str(e)}")
            if answer_data is not None:
                column_data["answer_value_text"] = str(answer_data)
        
        return column_data
    
    def upsert_answers_batch(
        self, 
        user_id: UUID, 
        payload: AnswerBatchUpdatePayload
    ) -> AnswerBatchResponse:
        """
        여러 답변을 배치로 UPSERT 처리합니다.
        
        Args:
            user_id: 사용자 UUID
            payload: 배치 답변 업데이트 요청 데이터
            
        Returns:
            AnswerBatchResponse: 처리 결과
        """
        processed_count = 0
        created_count = 0
        updated_count = 0
        failed_count = 0
        failed_requirements = []
        
        logger.info(f"답변 배치 처리 시작: user_id={user_id}, 답변 수={len(payload.answers)}")
        
        try:
            for answer_payload in payload.answers:
                try:
                    requirement_id = answer_payload.requirement_id
                    answer_data = answer_payload.answer_data
                    
                    # 요구사항의 데이터 타입 조회 (DisclosureRepository 사용)
                    requirement = self.disclosure_repository.get_requirement_by_id(requirement_id)
                    if not requirement:
                        logger.error(f"요구사항을 찾을 수 없음: {requirement_id}")
                        failed_count += 1
                        failed_requirements.append(requirement_id)
                        continue
                    
                    data_required_type = requirement.data_required_type
                    logger.debug(f"요구사항 {requirement_id}의 데이터 타입: {data_required_type}")
                    
                    # 기존 답변 존재 여부 확인 (생성/수정 카운트를 위해)
                    existing_answer = self.answer_repository.get_answer_by_user_and_requirement(
                        user_id, requirement_id
                    )
                    is_update = existing_answer is not None
                    
                    # 답변 데이터를 적절한 컬럼에 매핑
                    column_data = self._map_answer_data_to_columns(answer_data, data_required_type)
                    
                    # UPSERT 수행
                    answer = self.answer_repository.upsert_answer(
                        user_id=user_id,
                        requirement_id=requirement_id,
                        data_to_update=column_data
                    )
                    
                    if answer:
                        processed_count += 1
                        if is_update:
                            updated_count += 1
                        else:
                            created_count += 1
                        logger.debug(f"답변 처리 성공: {requirement_id} ({'업데이트' if is_update else '생성'})")
                    else:
                        failed_count += 1
                        failed_requirements.append(requirement_id)
                        logger.error(f"답변 처리 실패: {requirement_id}")
                        
                except Exception as e:
                    failed_count += 1
                    failed_requirements.append(answer_payload.requirement_id)
                    logger.error(f"답변 처리 중 예외 발생: {answer_payload.requirement_id}, error={str(e)}")
            
            # 트랜잭션 커밋
            self.db.commit()
            logger.info(f"답변 배치 처리 완료 및 커밋: user_id={user_id}, 성공={processed_count}, 실패={failed_count}")
            
        except Exception as e:
            # 트랜잭션 롤백
            self.db.rollback()
            logger.error(f"답변 배치 처리 중 심각한 오류 발생, 롤백: user_id={user_id}, error={str(e)}")
            # 전체 실패로 처리
            failed_count = len(payload.answers)
            failed_requirements = [answer.requirement_id for answer in payload.answers]
            processed_count = 0
            created_count = 0
            updated_count = 0
        
        success = failed_count == 0
        message = f"답변 배치 처리가 완료되었습니다. (성공: {processed_count}, 실패: {failed_count})"
        
        return AnswerBatchResponse(
            success=success,
            message=message,
            processed_count=processed_count,
            created_count=created_count,
            updated_count=updated_count,
            failed_count=failed_count,
            failed_requirements=failed_requirements
        )
    
    def get_answers_by_user_id(self, user_id: UUID) -> List[AnswerResponse]:
        """
        사용자의 모든 답변을 조회합니다.
        
        Args:
            user_id: 사용자 UUID
            
        Returns:
            List[AnswerResponse]: 사용자 답변 목록
        """
        try:
            # Repository에서 ORM 객체들을 받아옴
            answers = self.answer_repository.get_answers_by_user_id(user_id)
            
            # ORM 객체를 DTO로 변환하여 반환
            return [AnswerResponse.from_attributes(answer) for answer in answers]
            
        except Exception as e:
            logger.error(f"사용자 답변 목록 조회 실패: user_id={user_id}, error={str(e)}")
            return []
    
    def get_answer_by_requirement(
        self, 
        user_id: UUID, 
        requirement_id: str
    ) -> Optional[AnswerResponse]:
        """
        특정 요구사항에 대한 사용자 답변을 조회합니다.
        
        Args:
            user_id: 사용자 UUID
            requirement_id: 요구사항 ID
            
        Returns:
            AnswerResponse 또는 None
        """
        try:
            # Repository에서 ORM 객체를 받아옴
            answer = self.answer_repository.get_answer_by_user_and_requirement(user_id, requirement_id)
            
            # ORM 객체를 DTO로 변환하여 반환
            if answer:
                return AnswerResponse.from_attributes(answer)
            return None
            
        except Exception as e:
            logger.error(f"답변 조회 실패: user_id={user_id}, requirement_id={requirement_id}, error={str(e)}")
            return None
    
    def delete_answer(self, user_id: UUID, requirement_id: str) -> bool:
        """
        특정 답변을 삭제합니다.
        
        Args:
            user_id: 사용자 UUID
            requirement_id: 요구사항 ID
            
        Returns:
            bool: 삭제 성공 여부
        """
        try:
            result = self.answer_repository.delete_answer(user_id, requirement_id)
            if result:
                self.db.commit()
                logger.info(f"답변 삭제 완료 및 커밋: user_id={user_id}, requirement_id={requirement_id}")
            return result
            
        except Exception as e:
            self.db.rollback()
            logger.error(f"답변 삭제 실패 및 롤백: user_id={user_id}, requirement_id={requirement_id}, error={str(e)}")
            return False 